### ISSUE TITLE
TSO DELETE does not work ?

### ISSUE THREAD

**Issue by RonH2O** [mgrossmann/brexx370#18]
Apologies for the rather wordy posting, I guess I find it difficult to describe symptoms succinctly!

I had assumed that most TSO commands of the OS/VS2 era would be accessible using ADDRESS TSO - the ones I had used successfully  included ALLOC, FREE, SUBMIT and ATTR. However when I tried to issue DELETE an abend S806 results. I have posted the REXX and the resulting screen display and would be interested as to whether you regard this as a bug. What I am really to achieve is to create some JCL on the fly and submit it for execution by allocating a temporary file, writing JCL into it, submitting it and then deleting the file. It could well be that there is a better way to achieve the desired result - if so I would be happy to hear how it should be done. Anyway... the REXX and the result.

 ```
000001 jclStmt.1 = '//IS49DBR  JOB (01),'EXECUTE',CLASS=A,MSGCLASS=H,NOTIFY=IS49D'
 000002 jclStmt.2 = '//BR14   EXEC  PGM=MYBR14'
 000003 jclStmt.3 = '//STEPLIB  DD  DISP=SHR,DSN=HERC04.PDS.LOAD'
 000004 jclStmt.4 = '//SYSPRINT DD  SYSOUT=*'
 000005 jclStmt.5 = '//'
 000006 jclStmt.0 = 5
 000007 ADDRESS TSO
 000008 /*'ATTR DCB BLKSIZE(80) LRECL(80) RECFM(F) DSORG(PS)'*/
 000009 'ALLOC FI(JCL) DS(TMP.JCL) CATALOG SPACE(2,1) TRACKS USING(DCB)'
 000010 if RC \= 0 then do
 000011     say 'Error allocating file JCL. RC =' RC
 000012     exit 16
 000013 end
 000014 ADDRESS SYSTEM
 000015 fHandle = open(JCL,'w')
 000016 do i = 1 to jclStmt.0
 000017   jclRec = left(jclStmt.i,80)
 000018   nc = write(fHandle,jclRec)
 000019 end
 000020 call close(fHandle)
 000021 ADDRESS TSO 'FREE FI(JCL)'
 000022 say 'free returned rc' rc
 000023 say 'submitting job ...'
 000024 ADDRESS TSO 'SUBMIT TMP.JCL'
 000025 say 'job submitted. RC =' RC
 000026 ADDRESS TSO 'DELETE TMP.JCL'
 000027 say 'delete returned rc' rc
 000028 ADDRESS SYSTEM
 000029 exit 0

```
```
 ------------------------------------------------------------  RFE TSO  ------------------------------------------------------------
2ENTER TSO COMMAND OR CLIST BELOW:

 ===> rexx filetst3



 free returned rc 0
 submitting job ...
 job submitted. RC = 0
 ***

```
looks good, and the submitted JCL runs OK - but when the enter key is pressed I see .....
```
 BREXXSTD ABENDED S806 --------------------------------------  RFE TSO  ------------------------------------------------------------
2ENTER TSO COMMAND OR CLIST BELOW:

 ===>

```
... and the dataset (IS49D.TMP.JCL) is still catalogued and has to be removed by a subsequent delete issued outside of REXX.

Kind Regards
Ron

**Comment by mgrossmann** [comment]
Can you take a look in to the console log? Which module could not be found (806)?

the following is working well for me:

```
"LISTDS '"DSNSRC"'"                                               
IF RC == 0 THEN DO                                                
  "DELETE '"DSNTEMP"' SCRATCH NONVSAM"                            
END                                                               
                                                                  
"FREE FILE(TMP)"                                                  
"FREE ATTR(@DECK)"                                                
"ATTR @DECK RECFM(F B) LRECL(80) BLKSIZE(19040) DSORG(PS)"        
"ALLOC FILE(TMP) NEW REU DSN('"DSNTEMP"') USING(@DECK)"           
                                                                  
IF RC <> 0 THEN DO                                                
  RETURN -1                                                       
END    
```

**Comment by mgrossmann** [comment]
I will test your script later, today.

**Comment by mgrossmann** [comment]
@RonH2O  I can reproduce the error now. Thank you for reporting.

```
 READY                                            
brexx - address tso 'DELETE TMP.JCL'              
 COMMAND ALLOWS NO PARAMETERS - PARAMETERS IGNORED
 BREXX    ENDED DUE TO ERROR+                     
 READY                                            
?                                                 
 SYSTEM ABEND CODE 806   REASON CODE 004          
 READY       
```

or

```
 READY
brexx -i                                          
say hello                                         
 >>> HELLO                                        
address tso                                       
delete 'TMP.JCL'                                  
 COMMAND ALLOWS NO PARAMETERS - PARAMETERS IGNORED
 BREXX    ENDED DUE TO ERROR+                     
 READY                                            
?                                                 
 SYSTEM ABEND CODE 806   REASON CODE 004          
 READY                                            
```

**Comment by RonH2O** [comment]
Hi Mike - good to hear that it was not be being stupid (at least I hope not!). If you think I can do anything to assist then please, don't hesitate to ask me. I'm extremely grateful for the work you have done enabling REXX in MVS 3.8. Great job !!

Kind regards
Ron

Sent from Outlook<http://aka.ms/weboutlook>
________________________________

**Comment by mgrossmann** [comment]
Hi Ron, Thank you so much for your kind words. But the truth is that there are a lot of people to thank in the background. Beside @Peter-Jacob and me also @Juergen-Git (the TK4- man), Jason Winter (the JCC man), Greg Price (the REVIEW and IMON man) and last but not least Wally McLaughlin (the ISPF men) helped again and again that it goes ahead. :) 

By the way, I am very happy that you are busy testing and also reporting the bugs here. Only this way we can put the whole thing on solid feet. Thanks also for that.

But back to the real problem I'm not sure yet how it can go on here. Most likely it's not a problem of the BREXX/370. The TSO command DEL/DELETE is an alias of IDCAM01. If I call IDCAM01 in the TSO, the S806 is also used. Maybe @Juergen-Git can bring light into darkness at this point. But, we are on it. 

```
DEL     1K 000120 001312 IDCAM01  01 FO EP0         RN RU
DELETE  1K 000120 001312 IDCAM01  01 FO EP0         RN RU
```

```
 READY                                                        
brexx - address tso 'DELETE '''MIG.TMP.JCL''' SCRATCH NONVSAM'
 COMMAND ALLOWS NO PARAMETERS - PARAMETERS IGNORED            
 BREXX    ENDED DUE TO ERROR+                                 
 READY                                                        
?                                                             
 SYSTEM ABEND CODE 806   REASON CODE 004                      
 READY   
DELETE 'MIG.TMP.JCL' SCRATCH NONVSAM
 ENTRY (A) MIG.TMP.JCL DELETED      
 READY      
```

```
#include <stdio.h>                                          
#include <stdlib.h>                                         
#include <process.h>                                        
                                                            
int main (int argc, char *argv[])                           
{                                                           
    int rc = 0;                                             
    rc = systemTSO("DELETE 'MIG.TMP.JCL' SCRATCH NONVSAM"); 
    printf("FOO> rc = %d", rc);                             
    return rc;                                              
}          
```   

```
 READY      
testtso                                  
 TESTTSO  ENDED DUE TO ERROR+            
 READY                                   
?                                        
 SYSTEM ABEND CODE 806   REASON CODE 004 
 READY   
idcam01                                 
 IDCAM01  ENDED DUE TO ERROR+           
 READY                                  
?                                       
 SYSTEM ABEND CODE 806   REASON CODE 004
 READY                                  
```                 

Mike

**Comment by mgrossmann** [comment]
After a lot of research and a lot of help of brilliant people, 
we finally managed to solve the problem.

```
free returned rc 0                                                      
submitting job ...                                                      
job submitted. RC = 0                                                                                                            
$HASP165 JOB 4609  MIGTMPE ENDED- MAX COND CODE 0000                    
ENTRY (A) MIG.TMP.JCL DELETED                                                                                                     
delete returned rc 0             
```                                       

tell me, if you need a new load module.

Mike

**Comment by Peter-Jacob** [comment]
Und was war die Lösung?

 

Von: Mike Großmann <notifications@github.com> 
Gesendet: Mittwoch, 27. Februar 2019 09:44
An: mgrossmann/brexx370 <brexx370@noreply.github.com>
Cc: Peter-Jacob <PeterJ88@t-online.de>; Mention <mention@noreply.github.com>
Betreff: Re: [mgrossmann/brexx370] TSO DELETE does not work ? (#18)

 

Closed #18 <https://github.com/mgrossmann/brexx370/issues/18> .

**Comment by mgrossmann** [comment]
The load module IDCAM01 has several aliases. The module itself checks its own name after the call. Depending on which alias was called, different actions are executed. The name of the originally called program is determined via the field ECTPCMD of the ECT. Our runtime environment has passed the original ECT without adapting the ECTPCMD field.
